### ISSUE TITLE
Validation loss?

### ISSUE THREAD

**Issue by vlahm** [neuralhydrology/neuralhydrology#51]
Hello. I'd like to plot validation loss alongside training loss, so that I can see if there's a divergence indicative of an unrepresentative training/val set. I can't find a way to extract the validation loss from the model output.

**Comment by kratzert** [comment]
Hi @vlahm 

there are multiple ways, how you can look at the validation metrics.

1) They are logged automatically in TensorBoard, if you set `log_tensorboard: True` in the config. This will create a tensorboard log-file that you can a) view using `tensorboard --logdir /path/to/rundir` and b) use to extract all data in e.g. a Jupyter Notebook manually.
2) Setting `save_validation_results: True` in the config will store all validation results to disk in the run directory, in the same format as the test-set results. If you use the most recent version of this library, this will create a pickle file with the simulated time series and a csv-file with all metric values that you specify under `metrics` in the config.

However, if you are only interested to monitor the performance during training, I kinda recommend using TensorBoard.

Does this answer your questions?

**Comment by vlahm** [comment]
Thanks @kratzert. I've set both of those config parameters, and I see the relevant validation _metrics_, but I'm not seeing validation loss. If `loss` and `metrics` are set to the same function, they can be directly compared, but aren't `loss` and `metrics` computed independently during validation?

**Comment by kratzert** [comment]
I see, actually, we never compute the validation _loss_. I think there are two reasons for it:

1. We were actually never interested in it, but rather in the hydrological metrics. For example the `NSELoss` is just a adapted version of the NSE that allows us to use something similar to the NSE for model training with random mini-batches. 
2. The _metrics_ during validation are computed slightly different than the _loss_ during training. The loss is computed from a random mini-batch, where each sample can come from a different basin. During the validation, we evaluate on basin after the other, compute the metrics per basin and then eventually average over the basins to report the mean/median validation metric value. Considering that basins can have a different number of samples, averaging over the basins would yield a different loss than what we would see during training.

That doesn't mean that it is not possible, just that we saw no need for it so far. But I agree that this could be a useful addition. I'll discuss this with the others and see how quick we can add this.

**Comment by vlahm** [comment]
This makes sense. Given that the validation loss would necessarily be a different quantity than the training loss, comparing them directly doesn't feel as necessary. Still, it can sometimes be helpful to make diagnostic decisions based on relative magnitudes and/or slopes of these losses. I imagine that would hold even if validation loss was computed as an average of averages, though there would certainly be skewed edge-cases.

**Comment by kratzert** [comment]
Hi @vlahm, can you check if this works for you now and report feedback?

**Comment by vlahm** [comment]
@kratzert brilliant! This is just what I was looking for. Many thanks!
